### PR TITLE
DBZ-581 Improve error handling for replicators

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.jboss.org/browse/DBZ-581

### CODE REVIEW

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java [hunk 1]
```diff
@@ -35,6 +35,8 @@
  */
 public class ConnectionContext implements AutoCloseable {
 
+    private static final int PAUSE_AFTER_ERROR_MS = 500;
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java]: Can you add a documentation for the reasoning for this?

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java [hunk 2]
```diff
@@ -264,6 +266,12 @@ public void execute(String desc, Consumer<MongoClient> operation) {
                     return;
                 } catch (Throwable t) {
                     errorHandler.accept(desc, t);
+                    try {
+                        Thread.sleep(PAUSE_AFTER_ERROR_MS);
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/ConnectionContext.java]: Let's use the parking stuff with `Duration` from the `Metronome` class.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java [hunk 3]
```diff
@@ -167,6 +171,7 @@ public void stop() {
         try {
             logger.info("Stopping MongoDB connector");
             this.config = null;
+            if (replicaSetMonitorExecutor != null) replicaSetMonitorExecutor.shutdown();
             try {
                 if ( this.monitorThread != null ) this.monitorThread.shutdown();
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java]: Is this still needed with the executor?

Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java]: Yes, it is. The thread task does not return from `run` method until it receive an external signal for graceful termination.

Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnector.java]: Understood, but the request above, `replicaSetMonitorExecutor.shutdown()` is this signal (as it's the executor running this thread). I think `ReplicaSetMonitorThread.shutdown()` should be removed and replaced by correct reaction to thread interruption in its `run()` method. I.e. if it detects an interruption, it should abort its run loop. Then we won't even have to keep a reference to `monitorThread` here.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java [hunk 4]
```diff
@@ -170,6 +171,9 @@ public void start(Map<String, String> props) {
 
     @Override
     public List<SourceRecord> poll() throws InterruptedException {
+        if (replicatorError != null) {
+            throw new ConnectException("Failing connector task, at least one of replicators has failed");
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java]: s/one of/one of the/

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java [hunk 5]
```diff
@@ -202,6 +206,11 @@ public void stop() {
         return replContext.configureLoggingContext(CONTEXT_NAME);
     }
 
+    private void failedReplicator(Throwable t) {
+        replicatorError = t;
+        stop();
```
Comment (gunnarmorling) [inline debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java]: The `stop()` call shouldn't be needed, it'll be triggered by KC once the exception has been re-thrown by `poll()`. While it likely won't do harm, I'd prefer to keep it as simple as possible for the sake of understandability of the overall flow.

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java (hunk 4 above)
Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java]: ok

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java (hunk 5 above)
Comment (jpechane) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java]: I tend to disagree here - i went throught the KC code and I believe `stop()` is not called after `poll()` fails but it is vice-versa - `poll()` terminates when the `stop()` is invoked. Could you please double check it?

**Comment by jpechane** [comment]
@gunnarmorling Fixed and rebased to master

File: debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java (hunk 5 above)
Comment (gunnarmorling) [reply debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbConnectorTask.java]: Indeed, good catch!

**Comment by gunnarmorling** [comment]
Rebased and applied. Thanks, @jpechane!

**Pull request closed without merging**